lib/open_project/text_formatting.rb

```diff
@@ -56,473 +44,122 @@ def truncate_lines(string, options = {})
-      # TODO: transform modifications into WikiFormatting Helper, or at least ask the helper if he wants his stuff to be modified
-      @parsed_headings = []
-      text = parse_non_pre_blocks(text) { |text|
-        [:execute_macros, :parse_inline_attachments, :parse_wiki_links, :parse_redmine_links, :parse_headings, :parse_relative_urls].each do |method_name|
-          send method_name, text, project, obj, attr, only_path, options
-        end
-      }
+      return text if text.blank? or options[:edit]
 
-      if @parsed_headings.any?
-        replace_toc(text, @parsed_headings)
-      end
+      fragment = PrivateMethods::do_process text, options, self
 
-      escape_non_macros(text)
-      text.html_safe
+      fragment.to_s.html_safe
     end
     deprecated_alias :textilizable, :format_text
     deprecated_alias :textilize,    :format_text
 
-    ##
-    # Escape double curly braces after macro expansion.
-    # This will avoid arbitrary angular expressions to be evaluated in
-    # formatted text marked html_safe.
-    def escape_non_macros(text)
-      text.gsub!(/\{\{(?! \$root\.DOUBLE_LEFT_CURLY_BRACE)/, '{{ $root.DOUBLE_LEFT_CURLY_BRACE }}')
+    def current_request
+      request
+    rescue
+      nil
     end
 
-    def parse_non_pre_blocks(text)
-      s = StringScanner.new(text)
-      tags = []
-      parsed = ''
-      while !s.eos?
-        s.scan(/(.*?)(<(\/)?(pre|code)(.*?)>|\z)/im)
-        text = s[1]
-        full_tag = s[2]
-        closing = s[3]
-        tag = s[4]
-        if tags.empty?
-          yield text
-        end
-        parsed << text
-        if tag
-          if closing
-            if tags.last == tag.downcase
-              tags.pop
-            end
-          else
-            tags << tag.downcase
+    # prevent the following private methods from leaking into the including context
+    module PrivateMethods
+      require 'open_project/text_formatting/internal/transformers'
+
+      def self.prepare_processors(view)
+        # the order of processors is important and must not be changed
+        processors = [
+          # must run before all other processors during the process stage
+          OpenProject::TextFormatting::Internal::Transformers::RedmineWikiTransformer.new(view),
+          OpenProject::TextFormatting::Internal::Transformers::RedmineLinkTransformer.new(view),
+          OpenProject::TextFormatting::Internal::Transformers::WikiLinkTransformer.new(view),
+          # must run after redmine*/wiki link transformers during the
+          # process stage
+          OpenProject::TextFormatting::Internal::Transformers::MacroTransformer.new(view),
+          # the below two must run after macro transformer during the
+          # post process stage
+          OpenProject::TextFormatting::Internal::Transformers::ImageAttachmentTransformer.new(view),
+          # WTF?: all occurrences of {{ have already been replaced by {{ $root... }} by whatever
```

Comment:
This is caused by `angular-xss` gem which patches ERB and CGI.escape to do this transformation from non `html_safe` output to receive this transformation.